app/src/main/java/bacon/john/bitcoinpricewatch/bitcoin/BitcoinCurrencyManager.java

```diff
@@ -0,0 +1,120 @@
+package bacon.john.bitcoinpricewatch.bitcoin;
+
+import org.json.JSONObject;
+
+import java.util.*;
+import org.json.*;
+import java.text.*;
+import io.realm.*;
+
+import bacon.john.bitcoinpricewatch.localstorage.LocalStorageService;
+
+/**
+ * @class BitcoinCurrencyManager
+ *
+ * Retrieves information about bitcoin in various currencies
+ */
+public class BitcoinCurrencyManager {
+    public interface CurrencyListCallback {
+        public void callback(List<BitcoinCurrencyModel> currencies, Exception error);
+    }
+    public interface CurrencyCallback {
+        public void callback(BitcoinCurrencyModel currency, Exception error);
+    }
+
+    private static BitcoinCurrencyManager sharedInstance;
+
+    public static BitcoinCurrencyManager sharedInstance() {
+        if (sharedInstance == null) {
+            sharedInstance = new BitcoinCurrencyManager();
+        }
+        return sharedInstance;
+    }
+
+    public List<BitcoinCurrencyModel> getCachedCurrencies() {
+        return LocalStorageService.sharedInstance().getRealm().where(BitcoinCurrencyModel.class).findAll();
+    }
+
+    public BitcoinCurrencyModel getCachedCurrency(String currencyCode) {
+        return LocalStorageService.sharedInstance().getRealm().
+                where(BitcoinCurrencyModel.class).
+                equalTo("code", currencyCode).findFirst();
+    }
+
+    public void getRemoteCurrencyCodes(final CurrencyListCallback callback) {
+        BitcoinCurrencyService.sharedInstance().fetchCurrencyCodes(new BitcoinCurrencyService.Callback() {
+            @Override
+            public void callback(final JSONObject response, Exception error) {
+                if (error != null) {
+                    callback.callback(null, error);
+                    return;
+                } else {
+                    LocalStorageService.sharedInstance().getRealm().executeTransaction(new Realm.Transaction() {
```

Comment:
ensure we do JSON parsing on a worker Realm thread so we don't eat up the UI thread's cycles